shared/constants/chat.js

```diff
@@ -1224,21 +1234,86 @@ function getDeletedMessageIDs(state: TypedState, convIDKey: ConversationIDKey):
   return state.entities.deletedIDs.get(convIDKey, I.Set())
 }
 
-function getMessageUpdates(
-  state: TypedState,
-  messageKey: MessageKey
-): I.OrderedSet<EditingMessage | UpdatingAttachment> {
+function getMessageUpdateKeys(state: TypedState, messageKey: MessageKey): I.OrderedSet<MessageKey> {
   const {conversationIDKey, messageID} = splitMessageIDKey(messageKey)
-  const updateKeys = conversationIDKey
+  return conversationIDKey
     ? state.entities.messageUpdates.getIn([conversationIDKey, String(messageID)], I.OrderedSet())
     : I.OrderedSet()
-  return updateKeys.map(k => state.entities.messages.get(k)).filter(Boolean)
+}
+
+function getTextMessageUpdates(
+  state: TypedState,
+  messageKey: MessageKey
+): {last: ?EditingMessage, count: number} {
+  const updateKeys = getMessageUpdateKeys(state, messageKey)
+  return updateKeys.reduce(
+    (ret, k) => {
+      const message = state.entities.messages.get(k)
+      if (message && message.type === 'Edit') {
+        return {last: message, count: ret.count + 1}
+      } else {
+        return ret
+      }
+    },
+    {last: null, count: 0}
+  )
+}
+
+function getAttachmentMessageUpdates(
+  state: TypedState,
+  messageKey: MessageKey
+): {last: ?UpdatingAttachment, count: number} {
+  const updateKeys = getMessageUpdateKeys(state, messageKey)
+  return updateKeys.reduce(
+    (ret, k) => {
+      const message = state.entities.messages.get(k)
+      if (message && message.type === 'UpdateAttachment') {
+        return {last: message, count: ret.count + 1}
+      } else {
+        return ret
+      }
+    },
+    {last: null, count: 0}
+  )
+}
+
+function getMessageUpdateCount(state: TypedState, messageType: MessageType, messageKey: MessageKey): number {
+  if (messageType === 'Text') {
+    return getTextMessageUpdates(state, messageKey).count
+  } else if (messageType === 'Attachment') {
+    return getAttachmentMessageUpdates(state, messageKey).count
+  } else {
+    return 0
+  }
 }
 
 function getMessageFromMessageKey(state: TypedState, messageKey: MessageKey): ?Message {
   const message = state.entities.messages.get(messageKey)
-  const messageUpdates = getMessageUpdates(state, messageKey)
-  return message ? applyMessageUpdates(message, messageUpdates) : null
+  if (!message) {
+    return null
+  }
+
+  if (message.type === 'Text') {
```

Comment:
We don't really need a reduce loop here, right, since only the last update applies?

Reply:
mmmmmm @MarcoPolo

Reply:
I guess? there's nothing else we need to keep track of besides message, mention, and channel mention?

Reply:
@mmaxim I think might know better

Reply:
I'm not sure what this function is supposed to be doing? Is it applying supersedes messages?

Reply:
Yeah. Previously, it was applying all of them sequentially, and now it's just applying the last one, so the question is whether that is correct.

Reply:
Yes that is correct.